pkg/serviceprovider/gitlab/gitlab.go

```diff
@@ -0,0 +1,279 @@
+	// TODO: figure out how to categorize internal visibility
+	if project.Visibility == gitlab.PrivateVisibility || project.Visibility == gitlab.InternalVisibility {
+		status.Accessibility = api.SPIAccessCheckAccessibilityPrivate
+	}
+
+	return status, nil
+}
+
+// Temp
+func (g *Gitlab) publicRepo(ctx context.Context, accessCheck *api.SPIAccessCheck) (bool, error) {
+	lg := log.FromContext(ctx)
+	req, reqErr := http.NewRequestWithContext(ctx, "GET", accessCheck.Spec.RepoUrl, nil)
+	if reqErr != nil {
+		lg.Error(reqErr, "failed to prepare request", "accessCheck", accessCheck.Spec)
+		return false, fmt.Errorf("error while constructing HTTP request for access check to %s: %w", accessCheck.Spec.RepoUrl, reqErr)
+	}
+
+	resp, err := g.httpClient.Do(req)
+	if err != nil {
+		lg.Error(err, "failed to request the repo", "repo", accessCheck.Spec.RepoUrl)
+		return false, fmt.Errorf("error performing HTTP request for access check to %v: %w", accessCheck.Spec.RepoUrl, err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode == http.StatusOK {
+		return true, nil
+	} else if resp.StatusCode == http.StatusNotFound {
+		return false, nil
+	} else {
+		lg.Info("unexpected return code for repo", "repo", accessCheck.Spec.RepoUrl, "code", resp.StatusCode)
+		return false, nil
+	}
+}
+
+// Temp
+func (g *Gitlab) parseGitlabRepoUrl(repoUrl string) (repoPath string, err error) {
+	if !strings.HasPrefix(repoUrl, g.GetBaseUrl()) {
+		return "", fmt.Errorf("%w: '%s'", notGitlabUrlError, repoUrl)
+	}
+	return strings.TrimPrefix(repoUrl, g.GetBaseUrl()), nil
+}
+
+func (g Gitlab) GetOAuthEndpoint() string {
+	return g.Configuration.BaseUrl + "/gitlab/authenticate"
+}
+
+func (g Gitlab) MapToken(ctx context.Context, binding *api.SPIAccessTokenBinding, token *api.SPIAccessToken, tokenData *api.Token) (serviceprovider.AccessTokenMapper, error) {
+	return serviceprovider.DefaultMapToken(token, tokenData), nil
+}
+
+func (g Gitlab) Validate(ctx context.Context, validated serviceprovider.Validated) (serviceprovider.ValidationResult, error) {
+	ret := serviceprovider.ValidationResult{}
+
+	for _, p := range validated.Permissions().Required {
+		switch p.Area {
+		case api.PermissionAreaRepository,
+			api.PermissionAreaRepositoryMetadata,
+			api.PermissionAreaRegistry:
+			continue
+		case api.PermissionAreaUser:
+			if p.Type.IsWrite() {
+				ret.ScopeValidation = append(ret.ScopeValidation, userWritePermissionsNotSupportedError)
+			}
+		default:
+			ret.ScopeValidation = append(ret.ScopeValidation, fmt.Errorf("%w: '%s'", unsupportedAreaError, p.Area))
+		}
+	}
+
+	for _, s := range validated.Permissions().AdditionalScopes {
+		if !IsValidScope(s) {
+			ret.ScopeValidation = append(ret.ScopeValidation, fmt.Errorf("%w: '%s'", unsupportedScopeError, s))
+		}
+	}
+
+	return ret, nil
+}
+
+type gitlabProbe struct{}
+
+var _ serviceprovider.Probe = (*gitlabProbe)(nil)
+
+func (p gitlabProbe) Examine(_ *http.Client, url string) (string, error) {
+	// TODO: improve logic
```

Comment:
Do you have some ideas how to probe the host to figure out if it is a gitlab server?

Reply:
1. just regex match - something like 
- gitlab.service.com = yes
- service.domain.com/gitlab-repo = false
2. make some request to the URL that should respond in gitlab specific way
3. try to match the host part to some gitlab record in user config
4. others...

Reply:
Yep, something like that... Have you already tried to find some kind of a gitlab-specific request like that?

Reply:
The agreed-upon solution is as follows: The probe should have access to the service provider configuration and the configuration from custom secrets as in https://github.com/redhat-appstudio/service-provider-integration-operator/pull/302. Then compare the base of repoUrl from biding to the baseUrls off all GitLab service provider configurations.